source/agora/network/Client.d

```diff
@@ -604,35 +604,51 @@ public class NetworkClient
         enum name = __traits(identifier, endpoint);
         alias T = ReturnType!(__traits(getMember, API, name));
 
-        foreach (idx; 1 .. this.max_retries + 1)
+        T onError ()
         {
-            foreach (conn; this.connections)
+            static if (DT == Throw.Yes)
+            {
+                this.exception.file = file;
+                this.exception.line = line;
+                throw this.exception;
+            }
+            else static if (!is(T == void))
+                return T.init;
+        }
+
+    RETRY: foreach (idx; 1 .. this.max_retries + 1)
+        {
+            // Clients without connections should not wait and instead error out
+            if (!this.connections.length)
+                return onError();
+
+            foreach (index, conn; this.connections)
             {
                 if (this.banman.isBanned(conn.address))
                 {
-                    log.dbg("Client.attemptRequest '{}' skipped as {} is banned: {}/{}", name, conn.address, idx, this.max_retries);
+                    this.connections.remove(index);
+                    log.warn("Removing banned address {} while performing {}, addresses left: {}",
+                             conn.address, name, this.addresses);
+                    continue RETRY;
```

Comment:
This fixes the problem only for this instance of the loop, but we are removing from a list that could potentially be traversed by multiple fibers at any given time.

Reply:
@Geod24 , added a new commit to fix this and rebased.

Reply:
we can squash if you like it.